File: ulib/js/TestLib.js
/* @flow */

import * as FStar_Buffer from "./FStar_Buffer";

export type uint8_p = FStar_Buffer.buffer<number>;

export let uint8_p_to_string = (b:FStar_Buffer.buffer<number>) : ((len:number) => string) => ((len:number) => {
    let res = "";
    for (let i = 0; i < len; i++){
        res = res.concat(b._content[i].toString());
    };
    return res;
});

export let compare_and_print = (txt:any) => ((reference:uint8_p) => ((produced:uint8_p) => ((len:number) => {
    let ref_string = uint8_p_to_string(reference)(len);
    let prod_string = uint8_p_to_string(produced)(len);
    console.log("[test] expected output is \n" + ref_string);
    console.log("[test] computed output is \n" + prod_string);
    for (let i = 0; i < len; i++){
        if (ref_string.charAt(i) != prod_string.charAt(i)){
            console.log("[test] reference and expected strings differ at byte " + i);
            throw "Bad crypto";   
        }
    };
    console.log("Good crypto");
})));

File: examples/low-level/crypto/ml/TestLib.ml
open FStar_Buffer

type uint8_p = FStar_UInt8.t FStar_Buffer.buffer

let uint8_p_to_string b len : string =
  let s = ref ""  in
  for i = 0 to len - 1 do
    s := !s ^ Printf.sprintf "%02x" (index b i)
  done;
  !s

let compare_and_print txt (reference:uint8_p) (produced:uint8_p) len =
  let ref_string = uint8_p_to_string reference len in
  let prod_string = uint8_p_to_string produced len in
  Printf.printf "[test] expected output is %s\n" ref_string;
  Printf.printf "[test] computed output is %s\n" prod_string;
  for i = 0 to len - 1 do
    if (String.get ref_string i <> String.get prod_string i) then (
      Printf.printf "[test] reference %s and expected %s differ at byte %d\n" ref_string prod_string i;
      failwith "Bad crypto")
  done

let unsafe_malloc len = create 0 len
